### PR TITLE
Database Performance Audit

### PR DESCRIPTION
**Pull request by errea**
# Description

The goal of this project is to solve some performance issues, first, we need to introduce those issues. In order to do that, you will populate your database with a significant number of data. This PR is to check if the audit performance has been performed.

## Demo Before Audit

email_first_audit
vets_first_audit
visit_first_audit

## Demo After Audit
Screen Shot 2021-09-24 at 3 18 35 PM
Screen Shot 2021-09-24 at 3 19 35 PM
Screen Shot 2021-09-24 at 1 55 14 PM

## Project requirements 
The following queries are taking too much time (1 sec = 1000ms can be considered as too much time for database query). Try them on your machine to confirm it:
 * `SELECT COUNT(*) FROM visits where animal_id = 4`;
* `SELECT * FROM visits where vet_id = 2`;
* `SELECT * FROM owners where email = 'owner_18327@mail.com'`;
* Use `EXPLAIN ANALYZE `on the previous queries to check what is happening. Take screenshots of them - they will be necessary later.
* Find a way to decrease the execution time of the first query. Look for hints in the previous lessons.
* Find a way to improve the execution time of the other two queries.
* While you are making changes, check if they help by running `EXPLAIN ANALYZE`. Once you succeed, take a screenshot of the `EXPLAIN ANALYZE` result showing that time actually decreased.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added changes made to the database schema to the schema.sql file and committed in a new pull request.
In your pull request description include screenshots with EXPLAIN ANALYZE results (before and after you improve the speed) for each of the 3 problematic queries.
- [x] You need to use pair programming for this project. 
- [x] I have added a data query screenshot

### CODE REVIEW

File: data_query.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+CREATE INDEX animal_idx ON visits (animal_id);
+CREATE INDEX vet_idx ON visits (vet_id);
+CREATE INDEX owners_idx ON owners (email);
```
Comment (nidalaa) [inline data_query.sql:3]: [OPTIONAL]

This is a nice-to-have a suggestion as it won't improve the query that was reported as problematic in the requirements for this project.

As email addresses are not case sensitive, we might need to look for the emails in a case insensitive way. There is an option to reflect that in the index by using:

```
 CREATE INDEX ON owners ((lower(email)));
```

File: data_query.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+CREATE INDEX animal_idx ON visits (animal_id);
```
Comment (nidalaa) [inline data_query.sql:1]: @errea good call with the indexes.  I am quite sure that it helped but...

We are really missing the screenshots `after` you fixed the performance issues as requested [here](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_performance_audit.md#:~:text=.%20Once%20you%20succeed%2C%20take%20a%20screenshot%20of%20the%20EXPLAIN%20ANALYZE%20result%20showing%20that%20time%20actually%20decreased.).

Please add them to this PR description.

**Review by nidalaa — CHANGES_REQUESTED** [review]
Hi @errea , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: data_query.sql (hunk 2 above)
Comment (errea) [reply data_query.sql:1]: Ok, seen, would fix that

**Comment by errea** [comment]
Screenshot updated @nidalaa, thank you

**Review by nidalaa — CHANGES_REQUESTED** [review]
Hi , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Thank you for adding the screenshots @errea ! I can see that the query planner used index for `SELECT * FROM owners where email = 'owner_18327@mail.com'` and the time decreased.

But for the other two - nothing has changed.

Can you please run again the index creation queries:

```
CREATE INDEX animal_idx ON visits (animal_id);
CREATE INDEX vet_idx ON visits (vet_id);
```

And then use explain analyze again to create new screenshots?



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by errea** [comment]
Hi 😃 @nidalaa, thank you for your kind professional reviews so far, I have also updated the screenshots as you requested me to.

**Review by nidalaa — APPROVED** [review]
Hi  @errea 

For some reason, the query planner still prefers sequential scan (can you see the `seq scan` in the `explain analyze` output?) and ignores the indexes. Investigating that goes beyond the scope of this project.


The main goal of this project was to learn how to use the `explain analyzer` as a tool and to see how indexes can help for the slow queries. You could observe it in the output for the query with the owner email as a condition.


As you double-checked if indexes are really added and still the query planner does not use them, we can stop here. However, if you are curious you are really welcome to continue the investigation :)

-----


As I can clearly see that you had some fun with the `explain analyze` and I am sure that in the future you will know how to start the investigation for some slow DB queries, it's time to approve this project :)

There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

------
_As described in the [Students Handbook](https://www.notion.so/microverse/Project-reviews-022203ea6dd944b296b005533a19dcbe#cb67391d09f54bd98d6fed6ba24a4074) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by errea** [comment]
> Hi @errea
> 
> For some reason, the query planner still prefers sequential scan (can you see the `seq scan` in the `explain analyze` output?) and ignores the indexes. Investigating that goes beyond the scope of this project.
> 
> The main goal of this project was to learn how to use the `explain analyzer` as a tool and to see how indexes can help for the slow queries. You could observe it in the output for the query with the owner email as a condition.
> 
> As you double-checked if indexes are really added and still the query planner does not use them, we can stop here. However, if you are curious you are really welcome to continue the investigation :)
> 
> As I can clearly see that you had some fun with the `explain analyze` and I am sure that in the future you will know how to start the investigation for some slow DB queries, it's time to approve this project :)
> 
> There is nothing else to say other than... it's time to merge it :shipit:
> Congratulations! 🎉
> 
> _As described in the [Students Handbook](https://www.notion.so/microverse/Project-reviews-022203ea6dd944b296b005533a19dcbe#cb67391d09f54bd98d6fed6ba24a4074) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Yeah, truly and indeed, I learned a lot and read a lot on this, I deleted the database thrice and did it over again and again, thank you for your review @nidalaa and I hope to be working with this even on a larger scale ASAP 😄

**Pull request merged**